src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -1333,28 +1330,42 @@ public void populateDynamicChoices(ItemsetBinding itemset, TreeReference curQRef
             }
         }
 
+        Vector<SelectChoice> choices = new Vector<>();
         for (int i = 0; i < matches.size(); i++) {
-            TreeReference item = matches.elementAt(i);
+            choices.addElement(buildSelectChoice(matches.elementAt(i), itemset, formInstance, i));
+        }
+        itemset.setChoices(choices);
+    }
 
-            String label = itemset.labelExpr.evalReadable(fi, new EvaluationContext(exprEvalContext, item));
-            String value = null;
-            TreeElement copyNode = null;
+    private SelectChoice buildSelectChoice(TreeReference choiceRef, ItemsetBinding itemset,
+                                           DataInstance formInstance, int index) {
+        String label = itemset.labelExpr.evalReadable(formInstance,
+                new EvaluationContext(exprEvalContext, choiceRef));
+        String value = null;
+        TreeElement copyNode = null;
+        if (itemset.copyMode) {
+            copyNode = this.getMainInstance().resolveReference(itemset.copyRef.contextualize(choiceRef));
+        }
+        if (itemset.valueRef != null) {
+            value = itemset.valueExpr.evalReadable(formInstance,
+                    new EvaluationContext(exprEvalContext, choiceRef));
+        }
 
-            if (itemset.copyMode) {
-                copyNode = this.getMainInstance().resolveReference(itemset.copyRef.contextualize(item));
-            }
-            if (itemset.valueRef != null) {
-                value = itemset.valueExpr.evalReadable(fi, new EvaluationContext(exprEvalContext, item));
-            }
-            SelectChoice choice = new SelectChoice(label, value != null ? value : "dynamic:" + i, itemset.labelIsItext);
-            choice.setIndex(i);
-            if (itemset.copyMode)
-                choice.copyNode = copyNode;
+        SelectChoice choice = new SelectChoice(label, value != null ? value : "dynamic:" + index,
+                itemset.labelIsItext);
+        choice.setIndex(index);
+
+        if (itemset.copyMode) {
+            choice.copyNode = copyNode;
+        }
```

Comment:
Can you confirm that everything in this file up to here is just refactor? The diffs came out surprisingly hard to read...